File: canvas/hooks/spawn_plan_viewer.py
#!/usr/bin/env python3
"""
Spawn plan viewer when Claude writes/edits a plan file.
This hook triggers on PostToolUse for Write|Edit operations.
"""

import json
import os
import subprocess
import sys
from pathlib import Path


def debug_log(message):
    """Log debug messages to temp file."""
    try:
        log_file = Path(os.environ.get("TEMP", "/tmp")) / "canvas-hook-debug.log"
        with open(log_file, "a") as f:
            f.write(f"{message}\n")
    except Exception:
        pass


def get_lock_file(session_id):
    """Get session-specific lock file path."""
    temp_dir = Path(os.environ.get("TEMP", os.environ.get("TMPDIR", "/tmp")))
    return temp_dir / f"claude-plan-viewer-{session_id}.lock"


def is_plan_file(file_path):
    """Check if the file is a plan file (.claude/plans/*.md)."""
    # Normalize path separators
    normalized = file_path.replace("\\", "/")
    return ".claude/plans/" in normalized and normalized.endswith(".md")


def main():
    # Read input from stdin
    try:
        raw_input = sys.stdin.read()
        input_data = json.loads(raw_input)
    except json.JSONDecodeError as e:
        debug_log(f"JSON decode error: {e}")
        sys.exit(0)  # Allow tool to proceed

    # Extract tool information
    tool_name = input_data.get("tool_name", "")
    tool_input = input_data.get("tool_input", {})
    session_id = input_data.get("session_id", "default")

    # Only handle Write and Edit tools
    if tool_name not in ["Write", "Edit"]:
        sys.exit(0)

    # Get file path
    file_path = tool_input.get("file_path", "")
    if not file_path:
        sys.exit(0)

    # Check if this is a plan file
    if not is_plan_file(file_path):
        sys.exit(0)

    # Check lock file to prevent multiple viewers
    lock_file = get_lock_file(session_id)
    if lock_file.exists():
        debug_log(f"Lock file exists, skipping: {lock_file}")
        sys.exit(0)

    # Create lock file
    try:
        lock_file.write_text(str(os.getpid()))
    except Exception as e:
        debug_log(f"Failed to create lock file: {e}")

    # Get plugin root
    plugin_root = os.environ.get("CLAUDE_PLUGIN_ROOT", "")
    if not plugin_root:
        debug_log("CLAUDE_PLUGIN_ROOT not set")
        sys.exit(0)

    # Extract plan name from filename
    plan_name = Path(file_path).stem

    # Spawn the plan viewer in background
    try:
        cli_path = Path(plugin_root) / "src" / "cli.ts"

        # Use subprocess to spawn in background
        if sys.platform == "win32":
            # Windows: use CREATE_NEW_PROCESS_GROUP and DETACHED_PROCESS
            DETACHED_PROCESS = 0x00000008
            CREATE_NEW_PROCESS_GROUP = 0x00000200
            subprocess.Popen(
                ["bun", "run", str(cli_path), "watch-plan", file_path, "--title", f"Plan: {plan_name}"],
                cwd=plugin_root,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
                creationflags=DETACHED_PROCESS | CREATE_NEW_PROCESS_GROUP
            )
        else:
            # Unix: use nohup-style background
            subprocess.Popen(
                ["bun", "run", str(cli_path), "watch-plan", file_path, "--title", f"Plan: {plan_name}"],
                cwd=plugin_root,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
                start_new_session=True
            )

        debug_log(f"Spawned plan viewer for: {file_path}")
    except Exception as e:
        debug_log(f"Failed to spawn plan viewer: {e}")

    sys.exit(0)


if __name__ == "__main__":
    main()

File: canvas/hooks/scripts/spawn-plan-viewer.ts
#!/usr/bin/env bun
/**
 * Spawn plan viewer when Claude writes to a plan file
 * This hook is triggered by PostToolUse on Write operations
 * Cross-platform: Works on Windows, macOS, and Linux
 */

import { spawn } from "child_process";
import { existsSync, writeFileSync, readFileSync } from "fs";
import { basename, dirname, join } from "path";

async function main() {
  // Read JSON input from stdin
  const chunks: Buffer[] = [];
  for await (const chunk of Bun.stdin.stream()) {
    chunks.push(chunk);
  }
  const input = JSON.parse(Buffer.concat(chunks).toString());

  // Extract file path from tool_input
  const filePath = input?.tool_input?.file_path;
  if (!filePath) {
    process.exit(0);
  }

  // Check if this is a plan file (matches .claude/plans/*.md pattern)
  const normalizedPath = filePath.replace(/\\/g, "/");
  if (!normalizedPath.includes(".claude/plans/")) {
    process.exit(0);
  }

  // Check if file is a markdown file
  if (!filePath.endsWith(".md")) {
    process.exit(0);
  }

  // Lock file to prevent multiple viewers for the same session
  const sessionId = input?.session_id || "default";
  const tempDir = process.env.TEMP || process.env.TMPDIR || "/tmp";
  const lockFile = join(tempDir, `claude-plan-viewer-${sessionId}.lock`);

  // Check if viewer is already running for this session
  if (existsSync(lockFile)) {
    try {
      const content = readFileSync(lockFile, "utf-8").trim();
      if (content) {
        // Lock exists, viewer might be running
        process.exit(0);
      }
    } catch {
      // Ignore read errors
    }
  }

  // Create lock file
  try {
    writeFileSync(lockFile, process.pid.toString());
  } catch {
    // Ignore write errors
  }

  // Extract plan title from filename
  const planName = basename(filePath, ".md");
  const planTitle = `Plan: ${planName}`;

  // Get plugin root directory
  const pluginRoot = process.env.CLAUDE_PLUGIN_ROOT ||
    dirname(dirname(dirname(import.meta.dir)));

  const cliPath = join(pluginRoot, "src", "cli.ts");

  // Spawn the plan viewer in the background
  try {
    const proc = spawn("bun", ["run", cliPath, "watch-plan", filePath, "--title", planTitle], {
      cwd: pluginRoot,
      detached: true,
      stdio: "ignore",
      shell: process.platform === "win32",
    });
    proc.unref();
  } catch (e) {
    // Silently fail - don't block Claude
  }

  process.exit(0);
}

main().catch(() => process.exit(0));
